### PR TITLE
Rework how diagnostic lints are stored.

### PR DESCRIPTION
**Pull request by nnethercote**
`Diagnostic::code` has the type `DiagnosticId`, which has `Error` and
`Lint` variants. Plus `Diagnostic::is_lint` is a bool, which should be
redundant w.r.t. `Diagnostic::code`.

Seems simple. Except it's possible for a lint to have an error code, in
which case its `code` field is recorded as `Error`, and `is_lint` is
required to indicate that it's a lint. This is what happens with
`derive(LintDiagnostic)` lints. Which means those lints don't have a
lint name or a `has_future_breakage` field because those are stored in
the `DiagnosticId::Lint`.

It's all a bit messy and confused and seems unintentional.

This commit:
- removes `DiagnosticId`;
- changes `Diagnostic::code` to `Option<String>`, which means both
  errors and lints can straightforwardly have an error code;
- changes `Diagnostic::is_lint` to `Option<IsLint>`, where `IsLint` is a
  new type containing a lint name and a `has_future_breakage` bool, so
  all lints can have those, error code or not.
  
r? @oli-obk

### CODE REVIEW

**Comment by nnethercote** [comment]
FWIW, there are only three `derive(LintDiagnostic)` diagnostics with a code:
- `UnknownLintFromCommandLine` (E0602)
- `BindingsWithVariantName` (E0170)
- `ReprConflictingLint` (E0566)

That's few enough that I'm wondering if they're necessary and/or legitimate.

File: compiler/rustc_errors/src/diagnostic.rs [hunk 1]
```diff
@@ -104,7 +104,7 @@ pub struct Diagnostic {
     pub(crate) level: Level,
 
     pub messages: Vec<(DiagnosticMessage, Style)>,
-    pub code: Option<DiagnosticId>,
+    pub code: Option<String>,
```
Comment (cjgillot) [inline compiler/rustc_errors/src/diagnostic.rs:107]: Is there a way to store an integer instead of a string here? An `ErrorCode` type?

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 2]
```diff
@@ -1382,7 +1382,7 @@ fn compare_number_of_generics<'tcx>(
                     kind = kind,
                 ),
             );
-            err.code(DiagnosticId::Error("E0049".into()));
+            err.code("E0049".into());
```
Comment (cjgillot) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:1385]: Drive-by: this should use the `error_code!` macro.

File: compiler/rustc_errors/src/diagnostic.rs [hunk 3]
```diff
@@ -146,13 +146,11 @@ impl fmt::Display for DiagnosticLocation {
...
-pub enum DiagnosticId {
-    Error(String),
-    Lint {
-        name: String,
-        /// Indicates whether this lint should show up in cargo's future breakage report.
-        has_future_breakage: bool,
-    },
+pub struct IsLint {
+    /// The lint name.
+    pub(crate) name: String,
```
Comment (cjgillot) [inline compiler/rustc_errors/src/diagnostic.rs:151]: Can this use the `&Lint`? Or is there an issue with crate dependency order?

File: compiler/rustc_errors/src/diagnostic.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_errors/src/diagnostic.rs:107]: Yes! You're going to love #119972, which I just finished :)

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:1385]: I fixed that in #119972.

File: compiler/rustc_errors/src/diagnostic.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_errors/src/diagnostic.rs:151]: Interesting idea. Crate dependencies shouldn't be a problem. This struct has two fields:
- `name`, which is equivalent to `Lint::lower_name()`.
- `has_future_breakage`, which is a combination of `Lint::future_incompatible`, `Lint::default_level`, and `sess.opts.unstable_opts.future_incompat_test`. Adding a `sess` argument to `Diagnostic::has_future_breakage()` is awkward.

So I think keeping `IsLint` is probably best.
